### PR TITLE
WebGLRenderLists: Fix the issue that memory for disposed objects is not released

### LINKED ISSUE #12447: Memory not released after dispose
##### Description of the problem 

The created objects (mesh/material/geometry) are not released, after disposed in Three.js R87, while all released in Three.js R86, with the same code.

	var container;

	var camera, scene, renderer;

	init();
	animate();

	function init() {
		container = document.createElement( 'div' );
		document.body.appendChild( container );

		camera = new THREE.PerspectiveCamera( 60, window.innerWidth / window.innerHeight, 1, 10000 );
		camera.position.z = 100;

		scene = new THREE.Scene();

		renderer = new THREE.WebGLRenderer();
		renderer.setPixelRatio( window.devicePixelRatio );
		renderer.setSize( window.innerWidth, window.innerHeight );
		container.appendChild( renderer.domElement );
	}

	function animate() {
		requestAnimationFrame( animate );
		renderer.render( scene, camera );
	}
	
	var planes = [];
	function testplanes(){
		for(var i=0; i<100; i++){
			var m = new THREE.Mesh( 	new THREE.PlaneBufferGeometry( 1, 1 ), 
										new THREE.MeshBasicMaterial( {color: 0x010101*i } ));
			m.position.x = -50+i;
			scene.add( m );
			planes.push(m);
		}
	}			

	function delplanes(){
		for(var i=0; i<100; i++){
			scene.remove(planes[i]);	
			planes[i].geometry.dispose();
			planes[i].material.dispose();
			planes[i] = null;
		}	
	}
	

Typing testplanes() in console, then typing delplanes();
100 Meshes, 100 MeshBasicMaterial, and 100 PlaneBufferGeometry, remain in memory,  with memory profiling of heap snapshot,  in Three.js R87.  While all released in Three.js R86. 

##### Three.js version

- [ ] Dev
- [x] r87
- [ ] ...

##### Browser

- [ ] All of them
- [x] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [ ] All of them
- [x] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by takahirox**
This PR fixes #12447 

Let me know if there's any special reasons that `renderItems` wasn't cleared in `init()`

### CODE REVIEW

**Review by Mugen87 — APPROVED** [review]

**Comment by aardgoose** [comment]
Hi,
RenderItems isn't cleared in init() because it acts as a cache of RenderItem objects rather than having repeated allocations for each frame (init is called every frame). This patch will cause excessive heap churn and GC.

**Comment by aardgoose** [comment]
I came across this issue hence renderLists is exposed by the renderer so you can do:

renderer.renderLists.dispose();

to drop all the references to objects, materials etc after removing a set of objects or finishing with a camera / scene combination.

**Comment by takahirox** [comment]
You mean, `array.length = 0` can allocate new memory?

And I don't think `renderer.renderLists.dispose();` is a good idea.  I don't want users to be aware of `renderer.renderLists` for object disposal.

**Comment by takahirox** [comment]
Ah, you meant allocates new memory here, don't you.

https://github.com/takahirox/three.js/blob/871abf5cd090a67c5373bbdbf67f4dd7532d77ba/src/renderers/webgl/WebGLRenderLists.js#L73-L82

I also need to update it not to allocate new memory each frame...

**Comment by takahirox** [comment]
Updated

**Comment by aardgoose** [comment]
Correct, that's the allocation the render lists caches.

Don't know what the performance hit is of clearing all the references, in most cases, with an unchanged number of render items between frames you could probably avoid most of this clearing by using renderListIndex as a high water mark from the previous frame and clear from renderListIndex to renderList.length - 1 : That is, the RenderItems that haven't been overwritten in the previous frame.

I think this would in the common case avoid most of the overhead?

The renderLists.dispose() function is the only answer in some cases where a camera/scene combination won't be reused, but yes it is a very blunt object and a bit ugly except in special cases ( I required it after a rtt pass.

**Comment by takahirox** [comment]
Updated.

I think having `WebGLRenderList.finish()` again like r86 and clearing reference in it would be more straightforward.

@mrdoob Let me know if you don't wanna have `WebGLRenderList.finish()` again.

**Comment by Mugen87** [comment]
We should merge this PR or a similar solution. The are scenarios where `WebGLRenderList` effectively avoids garbage collection, see https://jsfiddle.net/v6bbxxa1/

Although 90 of the 100 objects are removed, the internal `renderItems` array of `WebGLRenderList` still has 100 entries with references to obsolete geometries.

**Comment by mrdoob** [comment]
This is how the code used to be before. I changed it because, depending on the scene, it would be creating and deleting objects all the time.

Would using `WeakMap`s help here?

**Comment by aardgoose** [comment]
@mrdoob 
You sadly couldn't replace the list with a WeakMap. You can't iterate them which is the price of weak references + keys must be objects, so the string 'hash' wouldn't work.

**Comment by anzorb** [comment]
Our application suffers from this in production and we are seeing GPU process crashes due to a memory leak. Running `renderer.renderLists.dispose()` makes the profile look clean again. How far is this PR from being complete/released and what do you recommend doing in the meantime? When is a good time to run `renderer.renderLists.dispose()`? Thanks!

**Comment by clanehin** [comment]
> This patch will cause excessive heap churn and GC.

> I changed it because, depending on the scene, it would be creating and deleting objects all the time.

I feel like minimizing GC activity can be a dubious tradeoff. If the objects being rendered onscreen are the only objects in the heap (and this is very possible now with web workers and web assembly), then the worst-case scenario of running the GC on every frame only visits the same objects that the renderer iterates every frame anyway.

That may be an unnecessary expense, but on the other hand, a memory leak that forces the tab to crash or the OS to swap (as I've been experiencing) is absolutely devastating to the user experience.

**Comment by Mugen87** [comment]
> This is how the code used to be before. I changed it because, depending on the scene, it would be creating and deleting objects all the time.

The new code is different than before. `finish()` does not invalidate render items anymore but only their references to other objects. In this way, the mentioned creation/deletion of objects should not happen. This is how `finish()` looked before:

https://github.com/mrdoob/three.js/blob/2667851c4a77355a5f46198cb5d4a16aebcdf80c/src/renderers/webgl/WebGLRenderLists.js#L117-L122

I think the approach of this PR is better than the current implementation since it prevents a clearly unexpected behavior of the library when disposing objects.

**Comment by Michalzr** [comment]
Hello,

I've been investigating a memory leak in our application that uses THREE.js. The leak wasn't solved even after I used the changes from this PR.

So what's the problem?
1) render the scene with camera A
2) dispose some of the used geometries
3) render the scene with camera B (and never use camera A again)
4) Memory leak -> the geometries are stored in render lists, and will never be removed

The obvious solution for me, is _not to use new instance of camera (or scene) once I used first one_.
So I'm going to do that. But I don't think this is intuitive for THREE.js user. Or is it?

Why are the render lists cached based on camera and scene id?

This is webgl_test_memory project, modified in a way that new instance of camera is created every frame.
Obviously this is not a real life example, it's just to prove my point - the memory grows rapidly there.
https://jsfiddle.net/x7Le8pda/

**Comment by Mugen87** [comment]
> Why are the render lists cached based on camera and scene id?

Because different view and projection matrices (derived from a camera) can lead to different depth values of objects and thus to a different sorting.

File: src/renderers/webgl/WebGLRenderLists.js [hunk 1]
```diff
@@ -130,6 +130,27 @@ function WebGLRenderList() {
...
+
+		for ( var i = renderItemsIndex, il = renderItems.length; i < il; i ++ ) {
+
+			var renderItem = renderItems[ i ];
+
+			if ( renderItem.id === null ) break;
+
+			renderItem.id = null;
+			renderItem.object = null;
+			renderItem.geometry = null
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLRenderLists.js]: Missing `;` 😇

**Comment by Mugen87** [comment]
There is already `Scene.dispose()` that can be used to free internal resources in `WebGLRenderLists` and `WebGLRenderStates`. The method is intended for the use case that you never want to use a scene again.

@ Michalzr We could actually add `Camera.dispose()` in order to delete the related render list and state. I guess this would solve your issue.

File: src/renderers/webgl/WebGLRenderLists.js (hunk 1 above)
Comment (takahirox) [reply src/renderers/webgl/WebGLRenderLists.js]: Updated. Thanks.

**Comment by Michalzr** [comment]
> We could actually add `Camera.dispose()` in order to delete the related render list and state. I guess this would solve your issue.

Having such method really seems useful 👍

>There is already `Scene.dispose()` that can be used to free internal resources in `WebGLRenderLists` and `WebGLRenderStates`. The method is intended for the use case that a scene objects becomes obsolete.

Yeah, you're right. I didn't notice before, thank you.

**Comment by Mugen87** [comment]
@Michalzr I've added a PR in order to address your issue.

**Comment by sgaoae** [comment]
Well, I think there might still be a problem. Consider this case: only one camera and one scene are used, and the user adds lots of memory-heavy objects for 2 times and then remove the objects from the latter time. These objects, however, will remain on memory as they are referenced by renderItems. 
And as the scene and the camera are always used, it's no that good to call either camera.dispose() or  scene.dispose(). 

What about exposing renderItems out so people can manage them freely? 
```
this.renderItems = renderItems
```
Or let people set a limit for the number of renderItems for init()?
```
let len = this.renderItems.length
if (len > limit) this.renderItems.length = limit
```
Or find a way to clear the renderItem for object when it is removed from scene?

**Comment by aardgoose** [comment]
in  WebGLRenderList.sort() you could do 

renderItems.length = transparent.length + opaque.length; to limit the cache to the current amount used.

Its the separation of cache between (scene, camera ) pairs that gives the bulk of the benefit here I think rather than caching withing a (scene, camera) pair, and would handle this case without setting arbitrary limits.

**Comment by sgaoae** [comment]
What about clear all of the renderItems after each render?
```
function clearRenderItem (renderItem) {
    renderItem.id = undefined;
    renderItem.object = undefined;
    renderItem.geometry = undefined;
    renderItem.material = undefined;
    renderItem.program = undefined;
    renderItem.groupOrder = undefined;
    renderItem.renderOrder = undefined;
    renderItem.z = undefined;
    renderItem.group = undefined;
}
renderList = renderer.renderLists.get(scene, camera)
for (let i = 0; i < renderList.renderItems.length; ++i) {
    clearRenderItem(renderList.renderItems[i])
}
// or clear opaque and transparent array instead?
for (let i = 0; i < renderList.opaque.length; ++i) {
    clearRenderItem(renderList.opaque[i])
}
for (let i = 0; i < renderList.transparent.length; ++i) {
    clearRenderItem(renderList.transparent[i])
}
```

**Comment by Mugen87** [comment]
@sgaoae I think the PR implemented this approach but in a slightly different way. Clearing the internal `renderItems` is sufficient since the same items of this array are in the opaque and transparent render lists.

**Comment by aardgoose** [comment]
@sgaoae if you have a case where the object count in a scene reduces significantly as you propose you can always use the brute force approach  which removes all the references, if this PR is not adopted.

renderer.renderLists.dispose();

**Comment by endel** [comment]
Just stumbled upon this issue by searching the web for "renderItems memory leak", as my app rebuilds the scene quite often, the suggestion @aardgoose made (https://github.com/mrdoob/three.js/pull/12464#issuecomment-338505684) seems perfect for me so far.

**Comment by viazmin** [comment]
I'm using a lot of large BufferGeometries and manually removing all the attributes from them before disposing, frees up most of the memory _(r104)_.

```javascript
for (const key in geometry.attributes) {
    geometry.removeAttribute(key);
}
geometry.setIndex([]);
geometry.dispose();
```

**Comment by Mugen87** [comment]
Closing in favor of #18411.

**Pull request closed without merging**